src/hyperdeck/response-handler.js

```diff
@@ -12,73 +12,78 @@ var ASYNC_RESPONSE_REGEX = /^(5[0-9]{2} )/;
-    lines.pop();
-    if (lines.length === 1) {
-      var asyncResponse = ASYNC_RESPONSE_REGEX.exec(lines[0]);
-      if (!asyncResponse) {
-        if (nextResponseIsMultiline) {
-          logger.debug('Assuming buffer is not complete because nextResponseIsMultiline flag is set.');
-          nextResponseIsMultiline = false;
-          return false;
-        }
-      }
-      // is it a single line response?
-      return lines[0].indexOf(":") !== lines[0].length-1;
-    } 
-    // multi line response so waiting for a blank line to signify end
-    return lines[lines.length-1] === "";
+  var buffer = [];
+
+  function isRespComplete() {
+    let complete = false;
+
+    if (buffer.length === 1) {
+      // a single line response
+      // single line response first line should never ends with ":" and
+      // 1XX and 200 is always single line response
+      complete = buffer[0].search(/^1\d\d\s|200\sok$/) === 0 && buffer[0].endsWith(':') === false;
+    } else if (buffer.length >= 3) {
+      // multi line response, so waiting for a blank line to signify end
+      complete = buffer[buffer.length - 1] === '';
+    }
+
+    return complete;
   }
-  
+
   function onData(rawData) {
-    logger.debug('Got data on socket.', rawData);
-    rawData.split("\r\n").forEach(function(line) {
-      if (buffer === '' && line.trim() === '') {
-        // handle empty lines before the data
-        // see https://github.com/LA1TV/Hyperdeck-JS-Lib/issues/44
-        return;
-      }
-      buffer += (line + "\r\n");
-      if (!isBufferComplete()) {
-        return;
+    logger.debug('Got data on socket.\n', rawData);
+
+    const resArray = rawData.split('\r\n');
+    // every response(even one line) must always end with \r\n, if not its not a valid response
+    if (resArray.length === 0 || resArray[resArray.length - 1] != '') {
+      throw "Unknown response type.";
+    }
+
+    resArray.pop();
+    resArray.forEach(function (line) {
+      // push to buffer till response is read completly
+      if (buffer.length > 0 || (buffer.length === 0 && line)) {
+        buffer.push(line);
       }
-      logger.debug('Got complete data.', buffer);
-      var data = Parser.parse(buffer);
-      // reset buffer
-      buffer = "";
-
-      switch (data.type) {
-        case "synchronousFailure":
-        case "synchronousSuccess":
-          var response = {
-            success: data.type === "synchronousSuccess",
-            data: data.data
-          };
-          notifier.emit("synchronousResponse", response);
-          break;
-        case "asynchronous":
-          notifier.emit("asynchronousResponse", data.data);
-          break;
-        default:
-          throw "Unknown response type.";
+
+      if (isRespComplete()) {
+        // remove the command seperator or extra last line
+        if (buffer[buffer.length - 1] === '') {
+          buffer.pop();
+        }
+
+        logger.debug('Got complete data.\n', buffer.join('\r\n'));
```

Comment:
Given it's logging I think
```javascript
logger.debug('Got complete data.', buffer);
```
is fine :)

Reply:
👍